runtime/gc_glue_java/ScavengerBackOutScanner.cpp

```diff
@@ -59,26 +59,25 @@ MM_ScavengerBackOutScanner::scanAllSlots(MM_EnvironmentBase *env)
 	/* Walk roots fixing up pointers through reverse forwarding information */
 	MM_RootScanner::scanAllSlots(env);
 
-	if (!_extensions->isConcurrentScavengerEnabled()) {
-		/* Back out Ownable Synchronizer and Continuation Processing */
-		MM_HeapRegionDescriptorStandard *region = NULL;
-		GC_HeapRegionIteratorStandard regionIterator(_extensions->heapRegionManager);
-		while (NULL != (region = regionIterator.nextRegion())) {
-			MM_HeapRegionDescriptorStandardExtension *regionExtension = MM_ConfigurationDelegate::getHeapRegionDescriptorStandardExtension(env, region);
-			for (uintptr_t i = 0; i < regionExtension->_maxListIndex; i++) {
-				/**
-				 * For Ownable Synchronizers lists, back out all of regions (includes tenure region, which is backed up during startProcessing).
-				 * Tenure list might have gotten new elements during main Scavenge phase and it is important to restore the list, since it will be iterated later during Marking Clearable phase.
-				 */
-				regionExtension->_ownableSynchronizerObjectLists[i].backoutList();
-				if ((MEMORY_TYPE_NEW == (region->getTypeFlags() & MEMORY_TYPE_NEW))) {
-					/**
-					 * For Continuation lists, Consistent with startProcessing that was earlier (in this GC cycle) called only on NEW regions.
-					 * Tenure list cannot get new elements, since they are added only during Scavenge Clearable phase that cannot be reached if there was a Scavenge abort.
-					 */
+	/* Back out Ownable Synchronizer and Continuation Processing */
+	MM_HeapRegionDescriptorStandard *region = NULL;
+	GC_HeapRegionIteratorStandard regionIterator(_extensions->heapRegionManager);
+
+	while (NULL != (region = regionIterator.nextRegion())) {
+		MM_HeapRegionDescriptorStandardExtension *regionExtension = MM_ConfigurationDelegate::getHeapRegionDescriptorStandardExtension(env, region);
+		if (_extensions->isConcurrentScavengerEnabled()) {
+			if (_scavenger->isObjectInEvacuateMemory((omrobjectptr_t )region->getLowAddress())) {
+				// for concurrent scavenger case, only backout lists in Evacuate region.
+				for (uintptr_t i = 0; i < regionExtension->_maxListIndex; i++) {
 					regionExtension->_continuationObjectLists[i].backoutList();
 				}
 			}
+		} else {
+			for (uintptr_t i = 0; i < regionExtension->_maxListIndex; i++) {
```

Comment:
we could pull out even the 2 for loops in front of 'if CS' check' 
it's a tradeoff between more code sharing vs performance (check CS for each sublist)
I slightly prefer simpler code, but again asking @dmitiri for an opinion

also, please convert the comments to c++ style

Reply:
For this case my vote is for more code sharing indeed. May be caching `_extensions->isConcurrentScavengerEnabled()` locally in the `const bool` variable can improve performance (but good compiler should do it already).

Reply:
@LinHu2016 sorry for all the confusion
more sharing would mean to push CS check more inside the loops (even inside the for loops)
but still keeping CS if check in front of other if checks for easier understanding
it is not optimal from perf point (to repeat same checks for same regions, while iterating through its sublists), but this is anyway very cheap processing, so we can live with it for the sake of code sharing and clarity

```
while regions
    for region lists 
       if CS
          if evac region then backout list
       else (if not CS)
          same as we used to do before the changes
```